### ISSUE TITLE
Put a limit on how complicated a complex (crypto-) condition can be

### ISSUE THREAD

**Issue by ttmc** [bigchaindb/bigchaindb#356]
The more complicated the condition, the more CPU work it takes to validate the fulfillment. A client could mess with BigchainDB by submitting transaction sequences with extremely complex conditions, tying up CPUs in validating the fulfillments. It's a nerdy and probably not very effective kind of denial-of-service attack.

I gather that there's an easy way to put a limit on how complex a condition (or fulfillment?) can be. Let's do that.

**Comment by libscott** [comment]
Also: Cap number of inputs a transaction may spend at once.

**Comment by libscott** [comment]
Strategy is to limit combined fulfillment size within a tx to some number ie 1kb.

**Comment by r-marques** [comment]
Limiting how complex a fulfillment can be would be easy. But in this case we would also need to the check the complexity of the conditions, or else we could allow for a condition that could never be fulfilled because its complexity would be higher than whatever we decided was the limit.

Another problem is that a transaction can have multiple fulfillments. In that case we would also need to limit the number of inputs of a transaction.

By limiting the number of inputs of a transaction we are now forcing users to create multiple transactions with a small number of inputs instead of a single transaction with a lot of inputs. In this case what would be the best? A transaction with 100 inputs or 100 transactions with 1 input?

In the end both the number of inputs and outputs as well as the complexity of the fulfillments all increase the size of the transactions so in the end by limiting the maximum size of a transaction we are already putting a limit on the number of inputs/outputs and complexity of transactions.

And size we can have/already have an imposed transaction size limit in the api I think that this sould be closed.

**Comment by ttmc** [comment]
Yes, one can put a limit on the size of requests sent in via the HTTP API. We already did that for the IPDB Test Network; we used an NGINX setting to limit the request size to a maximum of 15 kB.

If you think that's enough, then maybe add some documentation about that idea. This bit of documentation should also be revised:

https://docs.bigchaindb.com/projects/server/en/master/data-models/inputs-outputs.html

> When one creates a condition, one can calculate its fulfillment length (e.g. 96). The more complex the condition, the larger its fulfillment length will be. A BigchainDB federation can put an upper limit on the allowed fulfillment length, as a way of capping the complexity of conditions (and the computing time required to validate them).

**Comment by TimDaub** [comment]
While going through the guarantees I just found this ticket.

> We already did that for the IPDB Test Network; we used an NGINX setting to limit the request size to a maximum of 15 kB. If you think that's enough, then maybe add some documentation about that idea.

#1322 just covered revising documentation at: https://docs.bigchaindb.com/projects/server/en/master/data-models/inputs-outputs.html, namely:

> When one creates a condition, one can calculate its fulfillment length (e.g. 96). The more complex the condition, the larger its fulfillment length will be. A BigchainDB federation can put an upper limit on the allowed fulfillment length, as a way of capping the complexity of conditions (and the computing time required to validate them).

Not adding documentation on how (specifically "how", e.g. using NGINX or how to configure cryptocondition's maximum fulfillment length) to configure BigchainDB and peripheral software to restrict:

- transactions of a certain size to be accepted
- fulfillments of a certain size to be accepted.

Hence I'm proposing to reopen this ticket, to add this respective documentation. OK?

**Comment by ttmc** [comment]
The way we do it today, in NGINX, is we have the line
```text
    client_max_body_size 15k;
```
in the `nginx.conf` configuration file, to restrict the allowed size of the HTTP request body. Today, that line can be found here (for example): https://github.com/bigchaindb/nginx_3scale/blob/master/nginx.conf.template#L149 

Note that it only limits the size of the HTTP request body. Today, the biggest HTTP request body would be the body of a `POST /transaction` request, with a BigchainDB transaction in the body. That means that we're effectively limiting the size of a submitted BigchainDB transaction to 15k.

NGINX docs about client_max_body_size: https://nginx.org/en/docs/http/ngx_http_core_module.html#client_max_body_size

**Comment by ttmc** [comment]
I just added some documentation about *how* to enforce a maximum transaction size, using a reverse proxy such as NGINX; see pull request #1399.

I don't think there *is* a way to set a maximum fulfillment length in BigchainDB today (e.g. on the command line or in the BigchainDB config settings), so there's nothing to document about that.

As @r-marques commented above, enforcing a max transaction size is probably enough by itself, because it indirectly limits the crypto-conditions complexity.

I added a note to the above-mentioned docs page https://docs.bigchaindb.com/projects/server/en/master/data-models/inputs-outputs.html to say:

"Note: At the time of writing, there was no configuration setting to set a maximum allowed fulfillment length, so the only real option was to set a maximum allowed transaction size."

I believe my pull request #1399 resolves this issue, as well as #805 and #1303.
